route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -41,190 +41,238 @@ namespace route_following_plugin
+            resp.new_plan.maneuvers.push_back(latest_maneuver_plan_[i]);
+            ++i;
         }
+        
         if(resp.new_plan.maneuvers.size() == 0)
         {
             ROS_WARN_STREAM("Cannot plan maneuver because no route is found");
         }
+        //update plan
+
+        //update start distance of first maneuver and time for all maneuvers
+        setManeuverStartDist(resp.new_plan.maneuvers.front(), current_downtrack);
+        //Update time progress for maneuvers
+        updateTimeProgress(resp.new_plan.maneuvers, ros::Time::now());
+        //update starting speed of first maneuver
+        updateStartingSpeed(resp.new_plan.maneuvers.front(), current_speed_);
+
         return true;
     }
-    void RouteFollowingPlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
-    {
-        pose_msg_ = geometry_msgs::PoseStamped(*msg.get());
+
+    void RouteFollowingPlugin::updateTimeProgress(std::vector<cav_msgs::Maneuver>& maneuvers, ros::Time start_time){
+        ros::Time time_progress = start_time;
+        ros::Time prev_time = time_progress;
+
+        for(auto maneuver : maneuvers){
+            ros::Duration maneuver_duration =GET_MANEUVER_PROPERTY(maneuver,end_time) - GET_MANEUVER_PROPERTY(maneuver, start_time);
+            time_progress += maneuver_duration;
+            switch(maneuver.type){
+                case cav_msgs::Maneuver::LANE_FOLLOWING :
+                maneuver.lane_following_maneuver.start_time = prev_time;
+                maneuver.lane_following_maneuver.end_time = time_progress;
+                break;
+            case cav_msgs::Maneuver::LANE_CHANGE :
+                maneuver.lane_change_maneuver.start_time = prev_time;
+                maneuver.lane_change_maneuver.end_time = time_progress;
+                break;
+            case cav_msgs::Maneuver::INTERSECTION_TRANSIT_STRAIGHT :
+                maneuver.intersection_transit_straight_maneuver.start_time = prev_time;
+                maneuver.intersection_transit_straight_maneuver.end_time = time_progress;
+                break;
+            case cav_msgs::Maneuver::INTERSECTION_TRANSIT_LEFT_TURN :
+                maneuver.intersection_transit_left_turn_maneuver.start_time = prev_time;
+                maneuver.intersection_transit_left_turn_maneuver.end_time = time_progress;
+                break;
+            case cav_msgs::Maneuver::INTERSECTION_TRANSIT_RIGHT_TURN :
+                maneuver.intersection_transit_right_turn_maneuver.start_time = prev_time;
+                maneuver.intersection_transit_right_turn_maneuver.end_time = time_progress;
+                break;
+            case cav_msgs::Maneuver::STOP_AND_WAIT :
+                maneuver.stop_and_wait_maneuver.start_time = prev_time;
+                maneuver.stop_and_wait_maneuver.start_time = time_progress;
+                break;
+            default :
+                ROS_DEBUG_STREAM("Invalid maneuver type, cannot update time progress for maneuver");
```

Comment:
This should be an exception as the output will be garbage after failing to set these fields. Same with the other two methods like this.